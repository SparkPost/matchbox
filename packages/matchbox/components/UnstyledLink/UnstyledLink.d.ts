export default UnstyledLink;
declare const UnstyledLink: any;
