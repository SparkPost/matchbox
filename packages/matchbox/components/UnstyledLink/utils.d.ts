export function linkFrom({ content, ...action }: {
    [x: string]: any;
    content: any;
}, key: any): any;
