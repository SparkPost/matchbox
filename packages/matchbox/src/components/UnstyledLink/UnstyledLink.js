import React, { Component } from 'react';
import PropTypes from 'prop-types';

class UnstyledLink extends Component {
  static displayName = 'UnstyledLink';

  static propTypes = {
    to: PropTypes.oneOfType([PropTypes.string, PropTypes.object]),
    external: PropTypes.bool,
    component: PropTypes.oneOfType([
      PropTypes.func,
      PropTypes.element
    ]),
    children: PropTypes.node
  }

  render() {
    const {
      children,
      to,
      Component,
      component,
      external,
      ...rest
    } = this.props;

    const WrapperComponent = component || Component;

    if (to && !WrapperComponent) {
      return (
        <a
          href={to}
          target={external ? '_blank' : ''}
          rel={external ? 'noopener noreferrer' : ''}
          {...rest}>
          {children}
        </a>
      );
    }

    if (WrapperComponent) {
      return <WrapperComponent to={to} {...rest} >{children}</WrapperComponent>;
    }

    return <a {...rest}>{children}</a>;
  }
}

export default UnstyledLink;
