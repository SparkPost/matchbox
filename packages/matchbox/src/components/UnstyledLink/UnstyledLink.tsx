import React from 'react';
import styled from 'styled-components';
import {
  color,
  ColorProps,
  space,
  SpaceProps,
  typography,
  TypographyProps,
  compose,
} from 'styled-system';
import type * as Types from '../../helpers/types';

const system = compose(color, space, typography);

const Styledlink = styled.a`
  ${system}
  ${(props) => {
    if (props.disabled) {
      return `
        &, &:visited {
          opacity: 0.6;
          color: ${props.theme.colors.gray['900']};
          pointer-events: none;
        }
    `;
    }
  }}
`;

export type BaseProps = React.ComponentPropsWithRef<'a'> &
  Types.LinkActionProps &
  ColorProps &
  SpaceProps &
  TypographyProps;

type PolymorphicUnstyledLink = Types.ForwardRefComponent<'a', BaseProps>;

/**
 * UnstyledLink component
 * @see https://design.sparkpost.com/components/unstyledlink/api
 */
const UnstyledLink = React.forwardRef(function UnstyledLink(props, ref) {
  const {
    children,
    to,
    title,
    Component,
    component,
    as,
    external,
    onClick,
    role,
    disabled,
    tabIndex,
    ...rest
  } = props;

  const WrapperComponent = as || component || Component;
  const linkTitle = external && !title ? 'Opens in a new tab' : title;
  const linkRole = role ? role : !!onClick ? 'button' : null;

  const disabledAttributes = {
    'aria-disabled': disabled,
    disabled,
    tabIndex: disabled ? '-1' : tabIndex,
    onClick: disabled ? () => false : onClick,
  };

  if (to && !WrapperComponent) {
    return (
      <Styledlink
        as="a"
        href={to}
        target={external ? '_blank' : ''}
        rel={external ? 'noopener noreferrer' : ''}
        title={linkTitle}
        onClick={onClick}
        role={role}
        ref={ref}
        {...rest}
        {...disabledAttributes}
      >
        {children}
      </Styledlink>
    );
  }

  if (WrapperComponent) {
    return (
      <Styledlink
        as={WrapperComponent}
        onClick={onClick}
        ref={ref}
        role={role}
        title={linkTitle}
        to={to}
        {...rest}
        {...disabledAttributes}
      >
        {children}
      </Styledlink>
    );
  }

  return (
    <Styledlink
      as="a"
      title={linkTitle}
      role={linkRole}
      onClick={onClick}
      ref={ref}
      {...rest}
      {...disabledAttributes}
    >
      {children}
    </Styledlink>
  );
}) as PolymorphicUnstyledLink;

UnstyledLink.displayName = 'UnstyledLink';
export default UnstyledLink;
