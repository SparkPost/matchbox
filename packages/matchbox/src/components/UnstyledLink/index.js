export * from './utils';
export { default as UnstyledLink } from './UnstyledLink';
