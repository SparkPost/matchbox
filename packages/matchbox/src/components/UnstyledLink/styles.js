export const text = props => {
  if (props.as === 'a') {
    return `text-decoration: underline;`;
  }
};
