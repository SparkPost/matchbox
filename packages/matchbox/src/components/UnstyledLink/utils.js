import React from 'react';
import UnstyledLink from './UnstyledLink';

export function linkFrom({ content, ...action }, key) {
  return (
    <UnstyledLink
      key={key}
      children={content}
      {...action}
    />
  );
}
