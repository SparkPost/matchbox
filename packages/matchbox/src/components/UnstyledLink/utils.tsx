import React from 'react';
import UnstyledLink from './UnstyledLink';
import type { UnstyledLinkBaseProps } from './UnstyledLink';

interface LinkFromTypes extends UnstyledLinkBaseProps {
  content?: React.ReactNode;
  children?: React.ReactNode;
}

export function linkFrom(
  { content, children, ...action }: LinkFromTypes,
  key?: React.Key,
): JSX.Element {
  return (
    <UnstyledLink key={key} {...action}>
      {content}
      {children}
    </UnstyledLink>
  );
}
