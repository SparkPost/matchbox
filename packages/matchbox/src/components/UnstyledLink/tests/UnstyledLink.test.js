import React from 'react';
import UnstyledLink from '../UnstyledLink';
import { shallow } from 'enzyme';
import cases from 'jest-in-case';

describe('UnstyledLink', () => {
  let wrapper;
  beforeEach(() => {
    wrapper = shallow(<UnstyledLink>Hola!</UnstyledLink>);
  });

  const testCases = [
    { name: 'external button', props: { external: true }},
    { name: 'to without component', props: { to: '/nocomp' }},
    { name: 'external to without component', props: { to: '/nocomp', external: true }},
    { name: 'to with component', props: { to: '/withcomp', component: jest.fn }}
  ];

  cases('renders link states', (opts) => {
    wrapper.setProps(opts.props);
    expect(wrapper).toMatchSnapshot();
  }, testCases);
});
