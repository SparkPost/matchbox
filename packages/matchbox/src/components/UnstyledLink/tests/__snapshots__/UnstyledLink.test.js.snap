// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UnstyledLink renders link states external button 1`] = `
<a>
  Hola!
</a>
`;

exports[`UnstyledLink renders link states external to without component 1`] = `
<a
  href="/nocomp"
  rel="noopener noreferrer"
  target="_blank"
>
  Hola!
</a>
`;

exports[`UnstyledLink renders link states to with component 1`] = `
<bound fn
  to="/withcomp"
>
  Hola!
</bound fn>
`;

exports[`UnstyledLink renders link states to without component 1`] = `
<a
  href="/nocomp"
  rel=""
  target=""
>
  Hola!
</a>
`;
